Implement RFC 3349, mixed utf8 literals

nnethercote opened the pull request:
RFC: https://github.com/rust-lang/rfcs/pull/3349
Tracking issue: https://github.com/rust-lang/rust/issues/116907

r? @ghost

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7633318943/job/20795292767) [(plain)](https://github.com/rust-lang/rust/commit/bc4ce4abdae948ea1fa2b5c54f1e09b1c490ad06/checks/20795292767/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_ca26d7d0-62bc-4841-802c-a15a6f6cd72e
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=3349-mixed-utf8-literals
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_ca26d7d0-62bc-4841-802c-a15a6f6cd72e
GITHUB_REF=refs/pull/120286/merge
GITHUB_REF_NAME=120286/merge
GITHUB_REF_PROTECTED=false
---
Removing intermediate container 1f4c1d4af7d7
 ---> 39c7354a1621
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 90c83817214e
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
  Downloading virtualenv-20.25.0-py3-none-any.whl (3.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 19.1 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.1.0-py3-none-any.whl (17 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 43.3 MB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.1.0 virtualenv-20.25.0
Removing intermediate container 942307b1e25a
 ---> e6e3fc120e01
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 09892f35ff6c
 ---> 09892f35ff6c
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 88bee2b7dda1
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container d775561c9a59
 ---> e30e4329da7a
Successfully built e30e4329da7a
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:e30e4329da7ae50419177f7661e83fb243dfcc40327c70cbafca267be1580a5b
Uploading finished image sha256:e30e4329da7ae50419177f7661e83fb243dfcc40327c70cbafca267be1580a5b to https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
e30e4329da7a   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
09892f35ff6c   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
e6e3fc120e01   3 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
90c83817214e   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
39c7354a1621   11 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      12 days ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      12 days ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      12 days ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7fae0ba3b650>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/139141a603ff222e98d20ef66f0e4eaedfc4b373e41d302ecff577f90ee0c6cf0cab092d18a8f6a073c88b362b9a680d3b2b6e5c828faaf25d1b853eabb78b1a
sha256:e30e4329da7ae50419177f7661e83fb243dfcc40327c70cbafca267be1580a5b
---
DirectMap4k:      169920 kB
DirectMap2M:     6121472 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/5d3d3479d774754856db2db3e439dfb88ef3c52f/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-5d3d3479d774754856db2db3e439dfb88ef3c52f-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 25.21s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_parse/src/lexer/mod.rs at line 9:
 use rustc_ast::util::unicode::contains_text_flow_control_chars;
 use rustc_errors::{error_code, Applicability, DiagCtxt, DiagnosticBuilder, StashKey};
 use rustc_lexer::unescape::{self, EscapeError, Mode, Rfc3349};
-use rustc_lexer::{Base, DocStyle, RawStrError, Cursor, LiteralKind};
+use rustc_lexer::{Base, Cursor, DocStyle, LiteralKind, RawStrError};
 use rustc_session::lint::builtin::{
     RUST_2021_PREFIXES_INCOMPATIBLE_SYNTAX, TEXT_DIRECTION_CODEPOINT_IN_COMMENT,
 };
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_smir/src/rustc_internal/pretty.rs" "/checkout/compiler/rustc_smir/src/rustc_internal/internal.rs" "/checkout/library/profiler_builtins/src/lib.rs" "/checkout/library/profiler_builtins/build.rs" "/checkout/compiler/rustc_parse/src/lexer/tokentrees.rs" "/checkout/compiler/rustc_parse/src/lexer/unicode_chars.rs" "/checkout/compiler/rustc_parse/src/lexer/mod.rs" "/checkout/compiler/rustc_smir/src/rustc_internal/mod.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Jan 23 23:55:10 UTC 2024
  network time: Tue, 23 Jan 2024 23:55:10 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

nnethercote committed:
Implement RFC 3349, mixed utf8 literals.

Specifically:
- Allow unicode chars in b"" and br"" literals. This is done by changing
  `Mode::allow_unicode_chars` to succeed on `ByteStr` and `RawByteStr`.
- Allow unicode escapes in b"" literals. This is done by changing
  `Mode::allow_unicode_escapes` to succeed on `ByteStr`.

Byte string literals can already have high bytes (`\x80`..`\xff`).
Because they now also support unicode chars, they can now be mixed utf8,
so we use `unescape_mixed`/`cook_mixed` instead of
`unescape_unicode`/`cook_unicode` to process them.

A new type `Rfc3349`, is used to implement the feature gating. Values of
that type are threaded through the unescaping code to track whether
rules from rfc3349 are required for unescaping to succeed.

Test changes:

- tests/ui/mixed-utf8-literals/basic.rs: new `check-pass` UI test with
  various literals exercising the new forms.

- tests/ui/attributes/key-value-non-ascii.rs: changed from a byte string
  literal to a byte literal; we just need some kind of problem with a
  literal to preserve the test's intent.

- tests/ui/parser/raw/raw-byte-string-literals.rs: moved the raw byte
  string literal with a non-ASCII char to `basic.rs`.

- tests/ui/parser/byte-string-literals.rs: similar.

- tests/ui/parser/issues/issue-23620-invalid-escapes.rs: moved one case
  fully to `basic.rs`, and one partially.

- tests/ui/parser/unicode-control-codepoints.rs: left the code
  unchanged, but the errors are now about mixed-utf8-literals being
  feature gated.

- tests/ui/suggestions/multibyte-escapes.rs: moved one case to
  `basic.rs`.

- compiler/rustc_lexer/src/unescape/tests.rs: various adjustments
  - two cases that previously failed now succeed
  - added some more cases for the newly supported syntax

I wasn't sure how to handle rust-analyzer in general, so I just allowed
mixed utf8 literals everywhere without complaint.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7636496777/job/20803640906) [(plain)](https://github.com/rust-lang/rust/commit/f2dee333120a2c39236e3d52ba9b0b1e49cbe345/checks/20803640906/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_48d4d8be-d1ee-4c89-b3c5-58ccfc964097
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=3349-mixed-utf8-literals
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_48d4d8be-d1ee-4c89-b3c5-58ccfc964097
GITHUB_REF=refs/pull/120286/merge
GITHUB_REF_NAME=120286/merge
GITHUB_REF_PROTECTED=false
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Jan 24 06:44:01 UTC 2024
  network time: Wed, 24 Jan 2024 06:44:01 GMT
  network time: Wed, 24 Jan 2024 06:44:01 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished release [optimized] target(s) in 1.36s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --no-default-features --mini-tests --std-tests`
Using system GCC
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
..............................F.

failures:

error: test failed, to rerun pass `-p rustc_lexer --lib`
---- unescape::tests::test_unescape_raw_byte_str stdout ----
thread 'unescape::tests::test_unescape_raw_byte_str' panicked at compiler/rustc_lexer/src/unescape/tests.rs:291:9:
  left: Used
 right: Unused
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

</details>

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 585f3139a049efbcc3f83b4c76752881c0d111e8 with merge b626f8d9c6f947aa83d4a7f690e96eff73339152...
<!-- homu: {"type":"TryBuildStarted","head_sha":"585f3139a049efbcc3f83b4c76752881c0d111e8","merge_sha":"b626f8d9c6f947aa83d4a7f690e96eff73339152"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7650260238/job/20847749895)
Build commit: b626f8d9c6f947aa83d4a7f690e96eff73339152 (`b626f8d9c6f947aa83d4a7f690e96eff73339152`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7650260238/job/20847749895"},"merge_sha":"b626f8d9c6f947aa83d4a7f690e96eff73339152"} -->

rust-timer commented:
Queued b626f8d9c6f947aa83d4a7f690e96eff73339152 with parent 68411c955458ffbc98649d6dd057c4be3b187f38, future [comparison URL](https://perf.rust-lang.org/compare.html?start=68411c955458ffbc98649d6dd057c4be3b187f38&end=b626f8d9c6f947aa83d4a7f690e96eff73339152).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

nnethercote commented:
No significant perf changes.

@rustbot label: +perf-regression-triaged

nnethercote commented:
This doesn't fully implement the RFC, because I think a couple of things in the RFC are a bad idea. Details in [this comment](https://github.com/rust-lang/rust/issues/116907#issuecomment-1853160194).

bors commented:
:umbrella: The latest upstream changes (presumably #120491) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Dylan-DPC commented:
@nnethercote any updates on this? thanks

nnethercote commented:
I [nominated](https://github.com/rust-lang/rust/issues/116907#issuecomment-1912743726) it for lang-team discussion six months ago, and haven't heard anything since.

apiraino commented:
I'll signal that this needs to be discussed by T-lang

@rustbot label +S-blocked
